### PR TITLE
Exit options for tutorials

### PR DESCRIPTION
**Pull request by jaakkonarhi**
# Description of the PR

+ Opens an exit dialog if you click the dimmed area or the Cancel Tutorial button in the progress dropdown.

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [x] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [x] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [x] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialAction.java [hunk 1]
```diff
@@ -136,6 +130,14 @@ public boolean confirmStart(@NotNull TutorialViewModel tutorialViewModel) {
         JOptionPane.OK_CANCEL_OPTION) == JOptionPane.OK_OPTION;
     }
 
+    @Override
+    public boolean confirmCancel(@NotNull TutorialViewModel tutorialViewModel) {
+      return JOptionPane.showConfirmDialog(null,
+        "Cancel tutorial?",
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialAction.java:136]: Put this string into a resource file

Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialAction.java:136]: Put this string into a resource file

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  public TutorialActionGroup(@NotNull MainViewModelProvider mainViewModelProvider) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.addAll(new StepGroup(),
+        new Separator(),
+        new CancelAction());
+  }
+
+  private class CancelAction extends DumbAwareAction {
+    public CancelAction() {
+      super("Cancel Tutorial");
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java:37]: Should this be in resources?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      return IntStream.range(1, tutorial.getTasksAmount() + 1).mapToObj(StepAction::new)
+              .toArray(AnAction[]::new);
+    }
+  }
+
+  private class StepAction extends DumbAwareToggleAction {
+    private final int index;
+
+    public StepAction(int index) {
+      super("Step " + index);
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java:68]: More resources

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      var tutorial = mainViewModelProvider.getMainViewModel(e.getProject()).tutorialViewModel.get();
+      if (tutorial != null) {
+        return tutorial.getCurrentTaskIndex() == index;
+      }
+      return false;
+    }
+
+    @Override
+    public void setSelected(@NotNull AnActionEvent e, boolean state) {
+      //TODO change current task
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java:83]: Is this unimplemented on purpose

File: src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/ComponentLocator.java [hunk 5]
```diff
@@ -21,21 +19,6 @@ public static boolean hasChildOfClass(@NotNull Container c, @NotNull String clas
...
-      return true;
-    }
-
-    if (actionProperty != null && actionProperty.getClass().getName().contains(classSubstring)) {
-      return true;
-    }
-
-    return false;
-  }
-
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/ComponentLocator.java:38]: Don't remove it, even if it's unused

**Review by Taikelenn — APPROVED** [review]

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java (hunk 4 above)
Comment (jaakkonarhi) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java:83]: It's implemented in a later pr

Comment (jaakkonarhi) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialActionGroup.java:83]: It's implemented in a later pr

File: src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/ComponentLocator.java (hunk 5 above)
Comment (jaakkonarhi) [reply src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/ComponentLocator.java:38]: Can you make a javadoc comment for it please

Comment (Taikelenn) [reply src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/ComponentLocator.java:38]: Ok just delete it for now

**Review by stellatsv — APPROVED** [review]
Works well!

**Pull request merged**